backend/document/domain/assembly_strategies/assembly_strategy_utils.py

```diff
@@ -348,3 +350,109 @@ def ensure_primary_usfm_books_for_different_languages_are_adjacent(
                 )
             )
         )
+
+
+# NOTE This method works, but often still leaves unbalanced columns. To
+# be more effective, the algorithm would have to work at the intra-verse
+# content level to determine the midpoint to split columnar data. The
+# difficulty with this though is that the data is HTML not simple
+# content. This would mean getting into parsing HTML which in turn means
+# more complexity. Instead we can use an alternative HTML to PDF engine,
+# in our case Weasyprint, that understands CSS columns whereas
+# WKHTMLTOPDF does not. Weasyprint recently got header and footer
+# capability which was the only reason we were not using it before. Also
+# WKHTMLTOPDF has just fallen out of maintenance at the start of 2023 so
+# it is an easy decision. I am retaining this code for now and its
+# commented out invocations in case we need later (for some yet unknown
+# reason). It will likely disappear in a future PR.
```

Comment:
Thanks for this explanation.  I had comments about the commented code but I've removed them.  :)